lib/rules/object-shorthand.js

```diff
@@ -117,11 +131,86 @@ module.exports = {
             return node.type === "Literal" && typeof node.value === "string";
         }
 
+        /**
+         * Determines if the property is a shorthand or not.
+         * @param {ASTNode} property Property AST node
+         * @returns {boolean} True if the property is considered shorthand, false if not.
+         * @private
+         **/
+        function isShorthand(property) {
+
+            // property.method is true when `{a(){}}`.
+            return (property.shorthand || property.method);
+        }
+
+        /**
+         * Determines if the property's key and method or value are named equally.
+         * @param {ASTNode} property Property AST node
+         * @returns {boolean} True if the key and value are named equally, false if not.
+         * @private
+         **/
+        function isRedudant(property) {
+            return (property.key && (
+
+                // A function expression
+                property.value && property.value.id && property.value.id.name === property.key.name ||
+
+                // A property
+                property.value && property.value.name === property.key.name
+            ));
+        }
+
+        /**
+         * Ensures that an object's properties are consistently shorthand, or not shorthand at all.
+         * @param   {ASTNode} node Property AST node
+         * @param   {boolean} checkRedundancy Whether to check longform redundancy
+         * @returns {void}
+         **/
+        function checkConsistency(node, checkRedundancy) {
+
+            // We are excluding getters and setters as they are considered neither longform nor shorthand.
+            var properties = node.properties.filter(isNotGetterOrSetter),
+                shorthandProperties,
+                canAlwaysUseShorthand;
+
+            // Do we still have properties left after filtering the getters and setters?
+            if (properties.length > 0) {
+                shorthandProperties = properties.filter(isShorthand);
+
+                // If we do not have an equal number of longform properties as
+                // shorthand properties, we are using the annotations inconsistently
+                if (shorthandProperties.length !== properties.length) {
+
+                    // We have at least 1 shorthand property
+                    if (shorthandProperties.length > 0) {
+                        context.report(node, "Inconsistently using longform and shortform syntax.");
+                    } else if (checkRedundancy) {
+
+                        // If all properties of the object contain a method or value with a name matching it's key,
+                        // all the keys are redudant.
+                        canAlwaysUseShorthand = properties.every(isRedudant);
+
+                        if (canAlwaysUseShorthand) {
+                            context.report(node, "Properties shouldn't be in longform as all keys are redundant.");
```

Comment:
We usually put messages in the format "Expected x" or "Expected but found y.", So maybe "Expected shorthand for all properties."?